File: Datastructures + Algorithms/Datastructures/DSU/DSU.cpp
#include<bits/stdc++.h>
using namespace std;

struct DSU {
	int n;
	vector<int> parent;
	vector<int> sizes;

	DSU() {}
	DSU(int n) {
		this->n = n;
		parent.resize(n + 1);
		iota(parent.begin(), parent.end(), 0);
		sizes.resize(n + 1);
		fill(sizes.begin(), sizes.end(), 1);
	}

	int find_parent(int x) {
		if (parent[x] == x) {
			return x;
		}
		return parent[x] = find_parent(parent[x]);
	}

	bool union_sets(int a, int b) {
		a = find_parent(a);
		b = find_parent(b);

		if (a != b) {
			if (sizes[a] < sizes[b]) {
				swap(a, b);
			}

			parent[b] = a;
			sizes[a] += sizes[b];

			return 1;
		}

		return 0;
	}
};

File: Datastructures + Algorithms/Datastructures/DSU/DSU.py
class Union_Find:
    def __init__(self, n):
        self.n = n
        self.parent = list(range(n+1))
        self.size = [1]*(n+1)

    def find_parent(self, a):
        if self.parent[a] == a:
            return a

        self.parent[a] = self.find_parent(self.parent[a])
        return self.parent[a]

    def union_sets(self, a, b):
        a = self.find_parent(a)
        b = self.find_parent(b)

        if a != b:
            if self.size[a] < self.size[b]:
                a,b = b,a

            self.parent[b] = a
            self.size[a] += self.size[b]